### PR TITLE
feature: (2.34) Add translations support for SystemSettings

### PR DESCRIPTION
**Pull request by huzvak**
DHIS2-5527

### CODE REVIEW

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemSettingController.java [hunk 1]
```diff
@@ -184,39 +250,42 @@ public void getSystemSettingsJsonP( @RequestParam( value = "key", required = fal
...
-        Map<String, Serializable> value;
-        if ( keys != null && !keys.isEmpty() )
+        Set<SettingKey> settingKeys;
+
+        if ( keys != null )
         {
-            value = systemSettingManager.getSystemSettings( keys );
+            keys.removeIf( systemSettingManager::isConfidential );
+            settingKeys = keys.stream().map( key -> SettingKey.getByName( key ) ).filter( settingKeyOpt -> settingKeyOpt.isPresent() )
+                .map( settingKeyOpt -> settingKeyOpt.get() ).collect( Collectors.toSet() );
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemSettingController.java]: if you want, you can shorten it even more:

```
settingKeys = keys.stream().map( SettingKey::getByName ).filter( Optional::isPresent ).map( Optional::get )
                .collect( Collectors.toSet() );
```

File: dhis-2/dhis-services/dhis-service-setting/src/main/java/org/hisp/dhis/setting/DefaultSystemSettingManager.java [hunk 2]
```diff
@@ -155,6 +155,32 @@ public void saveSystemSetting( SettingKey key, Serializable value )
...
+        }
+        else
+        {
+            setting.getTranslations().put( locale, translation );
+        }
+
+        settingCache.invalidate( key.getName() );
+        systemSettingStore.update( setting );
+
+        return "";
```
[the remark this answers is not in the document]
Comment (huzvak) [reply dhis-2/dhis-services/dhis-service-setting/src/main/java/org/hisp/dhis/setting/DefaultSystemSettingManager.java]: Agree. I am not a fan myself. Somehow went under my radar, or was distracted by other details. Fixed.

File: dhis-2/dhis-services/dhis-service-setting/src/main/java/org/hisp/dhis/setting/DefaultSystemSettingManager.java [hunk 3]
```diff
@@ -229,6 +255,18 @@ public Serializable getSystemSetting( SettingKey key, Serializable defaultValue
...
+    @Override
+    public String getSystemSettingTranslation( SettingKey key, String locale )
+    {
+        SystemSetting setting = transactionTemplate.execute( status -> systemSettingStore.getByName( key.getName() ) );
+        if ( setting != null )
+        {
+            return setting.getTranslation( locale );
+        }
+
+        return "";
```
[the remark this answers is not in the document]
Comment (huzvak) [reply dhis-2/dhis-services/dhis-service-setting/src/main/java/org/hisp/dhis/setting/DefaultSystemSettingManager.java]: Agree. Fixed.

File: dhis-2/dhis-services/dhis-service-setting/src/main/java/org/hisp/dhis/setting/SystemSetting.java [hunk 4]
```diff
@@ -40,9 +42,8 @@
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.exc.MismatchedInputException;
 
+
 /**
- * TODO make IdentifiableObject
- * 
  * @author Stian Strandli
  */
 public class SystemSetting
```
[the remark this answers is not in the document]
Comment (huzvak) [reply dhis-2/dhis-services/dhis-service-setting/src/main/java/org/hisp/dhis/setting/SystemSetting.java]: Will leave it as it is. There are some specific get's, also from the documentation I understand that all attributes will be used in toString, equals, hashCode. So most probably also the ObjectMapper will be used. Etc. So leaving it as it is.

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemSettingController.java [hunk 5]
```diff
@@ -134,45 +179,66 @@ public void setSystemSettingV29( @RequestBody Map<String, Object> settings, Http
...
-        else
+        if ( !systemSettingManager.isConfidential( key ) )
         {
             Optional<SettingKey> settingKey = SettingKey.getByName( key );
 
-            Serializable setting;
-
             if ( settingKey.isPresent() )
             {
-                setting = systemSettingManager.getSystemSetting( settingKey.get() );
```
[the remark this answers is not in the document]
Comment (huzvak) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemSettingController.java]: Here it is a trade-off between nested IF statements and fetching data that are not necessary. Where fetching the data can be heavy. But I changed the code and eliminated nested if statements. If we will see that it is too heavy, we can introduce them back.

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemSettingController.java (hunk 1 above)
Comment (huzvak) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemSettingController.java]: I somehow missed complaint from Intellij and only copy-pasted the code. Fixed now.

File: dhis-2/dhis-services/dhis-service-setting/src/main/java/org/hisp/dhis/setting/SystemSetting.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (luciano-fiandesio) [reply dhis-2/dhis-services/dhis-service-setting/src/main/java/org/hisp/dhis/setting/SystemSetting.java]: ok

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemSettingController.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (luciano-fiandesio) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemSettingController.java]: I see your point and agree

**Pull request merged**